router/router.go

```diff
@@ -411,14 +415,46 @@ func (worker *workerT) workerProcess() {
 						WorkspaceId:   job.WorkspaceId,
 					}
 					worker.rt.responseQ <- jobResponseT{status: &status, worker: worker, userID: userID, JobT: job}
-					worker.rt.logger.Debugf(`Decrementing in throttle map for destination:%s since job:%d is marked as waiting for user:%s`, parameters.DestinationID, job.JobID, userID)
-					worker.rt.throttler.Dec(parameters.DestinationID, userID, 1, worker.throttledAtTime, throttler.ALL_LEVELS)
 					continue
 				}
 			}
 
-			firstAttemptedAt := gjson.GetBytes(job.LastJobStatus.ErrorResponse, "firstAttemptedAt").Str
+			var (
+				err            error
+				limited        bool
+				throttlingCost int64
+				tokensReturner throttling.TokenReturner
+				throttler      = worker.rt.getThrottler(parameters.DestinationID)
+			)
+			if len(worker.routerJobs) > 0 {
+				throttlingCost = worker.getThrottlingCost()
+			}
+			if throttlingCost > 0 && throttler != nil {
+				limited, tokensReturner, err = throttler.CheckLimitReached(userID, throttlingCost)
```

Comment:
Yes, bold move indeed. That's the thing that I got left basically, increase coverage in `router_test.go` and manual tests on dev to understand what can go wrong here. I'm pretty confident about the rest, not so much about the router changes atm.

Reply:
It is important to answer (and agree) why we need to apply throttling at the router, e.g. instead of relying to the destinations responding with a `HTTP 429`.

IMO, two of the most important reasons are:
1. Try to keep our workers busy with jobs that are most likely to succeed.
2. Avoid recording too many error statuses in jobsdb, whenever a job is rate limited.

The second reason might be more important than the first one, since the side-effects of having too many job statuses recorded in our jobsdb tables are multiple (and cascading):
- queries keep getting slower and slower
- the migrator is constantly busy migrating large tables in an attempt to lighten up the bloated status tables
- the backup loop is constantly busy trying to backup the never ending stream of migrated tables, it usually cannot keep up, causing a table pileup and eventually free disk space gets exhausted.

By moving the throttling gate out of the `generatorLoop` and inside each worker, I believe we are effectively cancelling both of the above points:
1. Jobs that might get rate limited will fill up our workers' queue
2. Two statuses will be recorded in jobsdb for every job that gets rate limited (`executing` & `failed`) which can end up getting bad pretty quickly.

Even though we are now trying to improve jobsdb's tolerance to bloated job status tables by introducing compaction, nevertheless it wouldn't be wise to consciously fill our job status tables with 429 errors.

Reply:
Makes sense, I'll revisit this and see what I can do.